File: tank_player_templates/tank_player_cpp/src/GameMap.cpp
/******************************************************************************
*
* Copyright Consoden AB, 2015
*
* Created by: Björn Weström / bjwe
*
*******************************************************************************/

#include <iostream>
#include <utility>
#include <vector>
#include <map>
#include <boost/date_time/posix_time/posix_time.hpp>

#include "GameMap.h"

GameMap::GameMap(int tankId, const Consoden::TankGame::GameStatePtr& gamePtr)
    :m_TankId(tankId)
    ,m_gamePtr(gamePtr)
    ,m_sizeX(gamePtr->Width().GetVal())
    ,m_sizeY(gamePtr->Height().GetVal())
{
    // Locate tanks
    Safir::Dob::Typesystem::ArrayIndex tankIndex;

    for (tankIndex = 0;
         tankIndex < m_gamePtr->TanksArraySize();
         tankIndex++) {

        if (m_gamePtr->Tanks()[tankIndex].IsNull()) {
            // empty tank slot, found last tank
            break;
        }

        Consoden::TankGame::TankPtr tankPtr =
                boost::dynamic_pointer_cast<Consoden::TankGame::Tank>(m_gamePtr->Tanks()[tankIndex].GetPtr());

        std::pair<int, int> pos = std::make_pair(tankPtr->PosX().GetVal(), tankPtr->PosY().GetVal());

        if (tankPtr->TankId().GetVal() == m_TankId) {
            // This is our tank!
            m_ownPos = pos;
        } else {
            // The enemy
            m_enemyPos = pos;
        }
    }
}

bool GameMap::IsWall(const std::pair<int, int>& pos) const
{
    return m_gamePtr->Board().GetVal()[Index(pos)]=='x';
}

bool GameMap::IsMine(const std::pair<int, int>& pos) const
{
    return m_gamePtr->Board().GetVal()[Index(pos)]=='o';
}

bool GameMap::IsCoin(const std::pair<int, int>& pos) const
{
    return m_gamePtr->Board().GetVal()[Index(pos)]=='$';
}

bool GameMap::IsPoisonGas(const std::pair<int, int>& pos) const
{
    return m_gamePtr->Board().GetVal()[Index(pos)]=='p';
}

bool GameMap::IsMissileInPosition(const std::pair<int, int>& pos) const
{
    Safir::Dob::Typesystem::ArrayIndex missileIndex;

    for (missileIndex = 0;
         missileIndex < m_gamePtr->MissilesArraySize();
         missileIndex++) {

        if (m_gamePtr->Missiles()[missileIndex].IsNull()) {
            // empty missile slot
            continue;
        }

        Consoden::TankGame::MissilePtr missilePtr =
                boost::dynamic_pointer_cast<Consoden::TankGame::Missile>(m_gamePtr->Missiles()[missileIndex].GetPtr());

        std::pair<int, int> head = std::make_pair(missilePtr->HeadPosX().GetVal(), missilePtr->HeadPosY().GetVal());
        std::pair<int, int> tail = std::make_pair(missilePtr->TailPosX().GetVal(), missilePtr->TailPosY().GetVal());

        if ((pos == head) || (pos == tail)) {
            return true;
        }
    }

    return false;
}

std::pair<int, int> GameMap::Move(const std::pair<int, int>& pos,
                                  Consoden::TankGame::Direction::Enumeration direction) const
{
    switch (direction)
    {
    case Consoden::TankGame::Direction::Left:
        return std::make_pair((pos.first - 1 + m_sizeX) % m_sizeX, pos.second);

    case Consoden::TankGame::Direction::Right:
        return std::make_pair((pos.first + 1) % m_sizeX, pos.second);

    case Consoden::TankGame::Direction::Up:
        return std::make_pair(pos.first, (pos.second - 1 + m_sizeY) % m_sizeY);

    case Consoden::TankGame::Direction::Down:
        return std::make_pair(pos.first, (pos.second + 1) % m_sizeY);

    case Consoden::TankGame::Direction::Neutral:
        return pos;
    }

    return pos;
}

int GameMap::TimeUntilNextJoystickReadout(int timestamp)
{
    // Get current time from the clock, using microseconds resolution
    const boost::posix_time::ptime now = 
        boost::posix_time::microsec_clock::local_time();

    // Get the time offset in current day
    const boost::posix_time::time_duration td = now.time_of_day();

    int total_milliseconds = td.total_milliseconds();

    // We ignore midnight to keep it simple
    return timestamp - total_milliseconds;
}    

void GameMap::PrintMap() const
{
    for (int y = 0; y < m_sizeY; y++) {
        for (int x = 0; x < m_sizeX; x++) {
            std::cout << m_gamePtr->Board().GetVal()[Index(x, y)];
        }
        std::cout << std::endl;
    }
    std::cout << std::endl;
}

void GameMap::PrintState() const
{
    std::cout << "Game board" << std::endl;
    PrintMap();

    std::cout << "Own position " << OwnPosition().first << "," << OwnPosition().second << std::endl;
    std::cout << "Enemy position  " << EnemyPosition().first << "," << EnemyPosition().second << std::endl;

    std::cout << "Active missiles" << std::endl;

    Safir::Dob::Typesystem::ArrayIndex missileIndex;

    for (missileIndex = 0;
         missileIndex < m_gamePtr->MissilesArraySize();
         missileIndex++) {

        if (m_gamePtr->Missiles()[missileIndex].IsNull()) {
            // empty missile slot
            continue;
        }

        Consoden::TankGame::MissilePtr missilePtr =
                boost::dynamic_pointer_cast<Consoden::TankGame::Missile>(m_gamePtr->Missiles()[missileIndex].GetPtr());

        std::pair<int, int> head = std::make_pair(missilePtr->HeadPosX().GetVal(), missilePtr->HeadPosY().GetVal());
        std::pair<int, int> tail = std::make_pair(missilePtr->TailPosX().GetVal(), missilePtr->TailPosY().GetVal());
        Consoden::TankGame::Direction::Enumeration direction = missilePtr->Direction().GetVal();

        std::cout << "Missile position - head " << head.first << "," << head.second << std::endl;
        std::cout << "Missile position - tail " << tail.first << "," << tail.second << std::endl;
        std::cout << "Missile direction ";

        switch (direction) {
        case Consoden::TankGame::Direction::Left:
            std::cout << "Left";
            break;
        case Consoden::TankGame::Direction::Right:
            std::cout << "Right";
            break;
        case Consoden::TankGame::Direction::Up:
            std::cout << "Up";
            break;
        case Consoden::TankGame::Direction::Down:
            std::cout << "Down";
            break;
        case Consoden::TankGame::Direction::Neutral:
            std::cout << "Neutral";
            break;
        }
        std::cout << std::endl;
    }

    std::cout << std::endl;
}

File: tank_player_templates/tank_player_java/src/cwg/GameMap.java
/******************************************************************************
*
* Copyright Consoden AB, 2014
*
* Created by: Joel Ottosson / joot
*
******************************************************************************/
package cwg;

import java.util.Calendar;

class GameMap {

	private long startTime;
	private int tankId;
	private consoden.tankgame.GameState gameState;
	private Calendar startOfDay;
	
	private byte mine=(byte)'o';
	private byte wall=(byte)'x';
	private byte coin=(byte)'$';
	private byte poison=(byte)'p';
	
	
	public GameMap(int tankId, consoden.tankgame.GameState gameState) {
		this.startTime=System.nanoTime();
		this.tankId=tankId;
		this.gameState=gameState;
		
		startOfDay=Calendar.getInstance();
		int year=startOfDay.get(Calendar.YEAR);
		int month=startOfDay.get(Calendar.MONTH);
		int date=startOfDay.get(Calendar.DATE);
		startOfDay.clear();
		startOfDay.set(year, month, date);
	}
	
	//X size of game field
	public int getSizeX() {
		return gameState.width().getVal();
	}
	
	//Y size of game field
	public int getSizeY() {
		return gameState.height().getVal();
	}
	
	//Own tank position
	public Position getOwnPosition() {
		return new Position(gameState.tanks().get(tankId).getObj().posX().getVal(),
				gameState.tanks().get(tankId).getObj().posY().getVal());
	}

	//Enemy tank position
	public Position getEnemyPosition() {
		int id=(tankId+1)%2;
		return new Position(gameState.tanks().get(id).getObj().posX().getVal(),
				gameState.tanks().get(id).getObj().posY().getVal());
	}
	
	//Check if square is a wall.
	public boolean isWall(final Position pos) {
		return gameState.board().getVal()[toIndex(pos)]==wall;
	}
	
	//Check if there is a mine in this square.
	public boolean isMine(final Position pos) {
		return gameState.board().getVal()[toIndex(pos)]==mine;
	}
	
	//Check if there is a coin in this square.
	public boolean isCoin(final Position pos) {
		return gameState.board().getVal()[toIndex(pos)]==coin;
	}
	
	//Check if there is poison gas in this square.
	public boolean isPoisonGas(final Position pos) {
		return gameState.board().getVal()[toIndex(pos)]==poison;
	}

	//Is there a missile in this square
	public boolean isMissileInPosition (Position p) {
		for (int i=0; i<this.gameState.missiles().size(); i++) {
			if (gameState.missiles().get(i).isNull())
				continue;

			consoden.tankgame.Missile missile = gameState.missiles().get(i).getObj();
			
			if ((p.x == missile.headPosX().getVal() && p.y == missile.headPosY().getVal()) ||
				(p.x == missile.tailPosX().getVal() && p.y == missile.tailPosY().getVal())) {
				return true;
			}
		}
		return false;
	}
	
	//Helper functions that moves position one step in the indicated direction.
	//Returns a new positon after the move operation.
	public Position move(final Position p, final consoden.tankgame.Direction d) {
		switch (d) {
        case LEFT:
        	return new Position((p.x-1+getSizeX())%getSizeX(), p.y);
        case RIGHT:
        	return new Position((p.x+1)%getSizeX(), p.y);
        case UP:
        	return new Position(p.x, (p.y-1+getSizeY())%getSizeY());
        case DOWN:
        	return new Position(p.x, (p.y+1)%getSizeY());
        case NEUTRAL:
        	return new Position(p.x, p.y);
        }
		return new Position(p.x, p.y);
	}
	
	//Milliseconds elapsed since creation of this instance
	public long getElapsedTime() {
		return (System.nanoTime()-startTime)/1000000;
	}
	
	public long timeToNextMove() {
		long time=Calendar.getInstance().getTimeInMillis()-startOfDay.getTimeInMillis();
		return gameState.nextMove().getVal().longValue()-time;
	}
	
	//Print game map
	public void printMap() {
		for (int y=0; y<getSizeY(); y++) {
			for (int x=0; x<getSizeX(); x++) {
				int i=toIndex(x, y);
				System.out.print((char)(gameState.board().getVal()[i]));
			}
			System.out.println();
		}
		System.out.println();
	}

	//Print game state
	public void printState() 
	{
	    System.out.println("Game board");
	    printMap();

	    System.out.println("Own position " + getOwnPosition());
	    System.out.println("Enemy position " + getEnemyPosition());

	    System.out.println("Active missiles");

		for (int i=0; i<this.gameState.missiles().size(); i++) {
			if (gameState.missiles().get(i).isNull())
				continue;

			consoden.tankgame.Missile missile = gameState.missiles().get(i).getObj();

			Position head = new Position(missile.headPosX().getVal(), missile.headPosY().getVal());
			Position tail = new Position(missile.tailPosX().getVal(), missile.tailPosY().getVal());
	        consoden.tankgame.Direction direction = missile.direction().getVal();

	        System.out.println("Missile position - head " + head);
	        System.out.println("Missile position - tail " + tail);
	        System.out.print("Missile direction ");

	        switch (direction) {
	        case LEFT:
	            System.out.println("Left");
	            break;
	        case RIGHT:
	            System.out.println("Right");
	            break;
	        case UP:
	            System.out.println("Up");
	            break;
	        case DOWN:
	            System.out.println("Down");
	            break;
	        case NEUTRAL:
	        	System.out.println("Neutral");
	            break;
	        }
	        System.out.println();
	    }

	    System.out.println();
	}

	private int toIndex(int x, int y) {
		return x+y*getSizeX();
	}
	
	private int toIndex(final Position pos) {
		return pos.x+pos.y*getSizeX();
	}
}
